### PR TITLE
IT-36770: Revise User Management and other Tickets

### PR DESCRIPTION
**Pull request by chhex**
Some changes in context of IT-36770. They enable an unintended setup which can be use for SNAPSHOT builds and the Build Pipeline. 

The important one in the following as overview:

- [x] Moved Hiera Data to separate dedicated repo - which will be established Apg internal
- [x] Established Hiera Hierarchie reflecting the test, prod und local inventory groups
- [x] For user jenkins and apg-patch-service-server ssh key are NOT generated anymore 
- [x] For user jenkins and apg-patch-service-server both private and public are retrieved and installed from Hiera in a seperate plan (accounts_sshkeys.pp)
- [x] Both jenkins and piper are configured to use the svcCvsClient 
- [x] Added a plan which configures remote usage of apscli for testing 

Next steps: 

- [x] Plan for setting deploy credentials with gradle-credentials-plugin
- [x] Create a specific user cmPiper for usage from piper 
- [ ] Create a group of Users apart from the --user <user> 
- [x] Move password and user configurations to the Hiera patchserver-setup-hiera repo depending on target group
- [x] Make user apg-patch-service-server and jenkins configuration data in hiera repo dependent of target group 
- [ ] Refactor artifactory user / profile configuration : conventions over configuration
- [x] Establish patchserver-setup-hiera apg git repo
- [x] Support transparent checkout patchserver-setup-hiera with sudo user 
- [x] Add Gradle credentials Task plan, see ticket IT-36839
- [x] Provide possibility to configure number of jenkins executors IT-36965

All the checked Task are tested. I am addressing also other Tickets, because this ticket is waiting for the new  apg git repository, before being able to merge.

### CODE REVIEW

File: hiera.yaml [hunk 1]
```diff
@@ -1,7 +1,9 @@
...
-  - name: Common
-    path: common.yaml
+  - name: "Common Data"
+    path: "common.yaml"
+  - name: "Environment"
+    path: "environment/%{::group}.yaml"
 defaults:
   data_hash: yaml_data
-  datadir: hiera-data
+  datadir: /Users/chhex/git/patchserver/patchserver-setup-hiera/data
```
Comment (chhex) [inline hiera.yaml:9]: This absolute will go away when the hiera repo can be pulled from a git server

File: inventory-local.yaml [hunk 2]
```diff
@@ -10,15 +10,14 @@ vars:
   piper_bind_address:  # empty for local testing
   piper_cvs_remote_host: cvs-t.apgsga.ch
   piper_spring_profiles: live,remotecvs, patchOMock
-  piper_cvs_user: che # TODO (che, 7.1) -> svcCvsClient
-  jenkins_ssh_user: che # TODO (che, 7.1)  -> apg-patch-service-server
-  install_ssh_user: # TODO (che, 07.1)  -> apg_install
+  jenkins_ssh_user: che # TODO (che, 7.1)  -> cm
```
Comment (chhex) [inline inventory-local.yaml:13]: Still pointing to the sudo user, but will be generic user in one of the next steps

File: inventory-local.yaml [hunk 3]
```diff
@@ -10,15 +10,14 @@ vars:
   piper_bind_address:  # empty for local testing
   piper_cvs_remote_host: cvs-t.apgsga.ch
   piper_spring_profiles: live,remotecvs, patchOMock
-  piper_cvs_user: che # TODO (che, 7.1) -> svcCvsClient
-  jenkins_ssh_user: che # TODO (che, 7.1)  -> apg-patch-service-server
```
Comment (chhex) [inline inventory-local.yaml:14]: Will probably move to the hiera repo

File: plans/accounts_sshkeys.pp [hunk 4]
```diff
@@ -0,0 +1,58 @@
+plan piper::accounts_sshkeys (
```
Comment (chhex) [inline plans/accounts_sshkeys.pp:1]: Plan, which manages ssh keys for jenkins and apg-patch-service-server

File: plans/jenkins_account_create.pp [hunk 5]
```diff
@@ -12,14 +12,12 @@
         # Default password , enrypted with openssl passwd -1 on target
         password => '$1$7vKrRq8U$K0dm4xWqhJ0M5tbvAmypl/',
       }
-      ssh_keygen { $user: }
```
Comment (chhex) [inline plans/jenkins_account_create.pp:15]: The ssh keys are NOT generated anymore , but installed , see new plan above

File: plans/jenkins_dirs_create.pp [hunk 6]
```diff
@@ -3,36 +3,21 @@
...
-  # apg-patch-service-server user on target
-  $dl_patch_result = download_file("${piper_ssh_dir}/id_rsa.pub", 'piper', $targets)
-  $patch_rsa_target = file::read($dl_patch_result.first['path'])
   # User public rsa key on target on Host
   # TODO (che, 19.11) is this disireable of production target
   $home_dir = system::env('HOME')
   $user_rsa_local = file::read("${$home_dir}/.ssh/id_rsa.pub")
-  $ssh_keys = "${user_rsa_target}\n${user_rsa_local}\n${patch_rsa_target}"
   apply($targets) {
+    # apg-patch-service-server user on target
```
Comment (chhex) [inline plans/jenkins_dirs_create.pp:18]: Read via Hiera Look-up

File: plans/piper_service_properties.pp [hunk 7]
```diff
@@ -15,7 +15,7 @@
           $result = download_file("/etc/opt/apg-patch-service-server/${type}_properties.initial", 'piper', $targets)
           out::message($result)
           $applyResult = apply($targets) {
-            $content = epp($result.first['path'], { 'piper_cvs_user' => "${user}", 'jenkins_ssh_user' => "${user}" })
+            $content = epp($result.first['path'], { 'piper_cvs_user' => "${targetall.vars[cvs_user]}", 'jenkins_ssh_user' => "${user}" })
```
Comment (chhex) [inline plans/piper_service_properties.pp:18]: Central cvs_user for jenkins and piper

File: plans/apscli_remote_test_config.pp [hunk 8]
```diff
@@ -0,0 +1,23 @@
+plan piper::apscli_remote_test_config(
```
Comment (chhex) [inline plans/apscli_remote_test_config.pp:1]: Plan for testing purposes , make it possible to use apscli remote, instead on the target

**Comment by chhex** [comment]
@apgsga-jhe  @apgsga-uge  This is work in progress  , but feedback / input very welcome. 
The functionality with this last commit have been locally tested.

**Comment by apgsga-jhe** [comment]
@chhex : Looking good to me. A very small question regarding the next steps, and "Create a specific user cmJenkinsUser for usage from piper": That will be the user called "cm" on Jenkins, right?
@apgsga-uge : FYI

**Review by apgsga-jhe — APPROVED** [review]

**Comment by chhex** [comment]
> ..A very small question regarding the next steps, and "Create a specific user cmJenkinsUser for usage from piper": That will be the user called "cm" on Jenkins, right?
Yep, what do you guys , prefer in terms of naming ?  For me <cm> is a little too generic in terms of Usecase. <cmJenkinsUser> for that too, how about <piperJenkinsUser> or <patchServiceJenkinsUser>... so that from the naming the useCase is clear. Otherwise the same can happen, as with <svcCvsClient> where we don't want to change the Password, because i's usage scenarios are not clear anymore.

**Comment by apgsga-jhe** [comment]
> > ..A very small question regarding the next steps, and "Create a specific user cmJenkinsUser for usage from piper": That will be the user called "cm" on Jenkins, right?
> 
> @apgsga-jhe Yep, what do you guys prefer in terms of naming ? For me is a little too generic in terms of Usecase. for that too, how about or ... so that from the naming the useCase is clear. Otherwise the same can happen, as with where we don't want to change the Password, because i's usage scenarios are not clear anymore.
> 
> @apgsga-uge

@chhex : "cm" would be ok for me
@apgsga-uge : what do you think?

**Comment by chhex** [comment]
@apgsga-jhe @apgsga-uge  Added Plan for setting deploy credentials with gradle-credentials-plugin.
See plans/gradle_cred_setup.pp

**Comment by chhex** [comment]
@apgsga-jhe @apgsga-uge  Added specific Jenkins User for Piper

**Comment by chhex** [comment]
@apgsga-jhe @apgsga-uge  Number of executors are now configurable (IT-36965) , see  commit 
09bdcda

File: Gemfile [hunk 9]
```diff
@@ -2,7 +2,7 @@ source 'https://rubygems.org'
 source 'https://artifactory4t4apgsga.jfrog.io/artifactory/api/gems/apgGems' do
   gem 'apgsecrets',  '~> 0.2.9', '>= 0.2.9'
   gem 'artcli' ,  '~> 0.9', '>= 0.9.3'
-  gem 'apscli' ,  '~> 0.8', '>= 0.8.0'
+  gem 'apscli' ,  '~> 0.8', '>= 0.8.1'
```
Comment (chhex) [inline Gemfile:5]: Upgraded ruby gem apscli wrapper with newest fat jar

File: inactive-plans/jenkins_gradle_home.pp [hunk 10]
```diff
@@ -0,0 +1,31 @@
+plan piper::jenkins_gradle_home (
```
Comment (chhex) [inline inactive-plans/jenkins_gradle_home.pp:1]: This is actually inactive and not used , see Ticket IT-36776 for reasons

**Comment by chhex** [comment]
@apgsga-jhe @apgsga-uge  Made user apg-patch-service-server and jenkins configuration data in hiera repo dependent of target group (development, integration, production). 
The target group config is also for the Yum artifactory configuration
See also commit 
a2594ea

**Comment by chhex** [comment]
@apgsga-uge  This repo is in sync with the apg internal repo.  I will merge, but keep the pull request for reference. 
@apgsga-jhe  FYI

**Pull request merged**